plugin/healthchecker/filter.go

```diff
@@ -0,0 +1,121 @@
+package healthchecker
+
+import (
+	"time"
+
+	"github.com/hashicorp/golang-lru"
+	"github.com/miekg/dns"
+	"go.uber.org/atomic"
+)
+
+type (
+	HealthCheckFilter struct {
+		cache    *lru.Cache
+		checker  Checker
+		interval time.Duration
+		names    map[string]struct{}
+	}
+
+	entry struct {
+		dnsRecord dns.RR
+		healthy   *atomic.Bool
+		quit      chan struct{}
+	}
+)
+
+func NewHealthCheckFilter(checker Checker, size int, interval time.Duration, names map[string]struct{}) (*HealthCheckFilter, error) {
+	cache, err := lru.NewWithEvict(size, func(key interface{}, value interface{}) {
+		if e, ok := value.(*entry); ok {
+			close(e.quit)
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &HealthCheckFilter{
+		cache:    cache,
+		checker:  checker,
+		interval: interval,
+		names:    names,
+	}, nil
+}
+
+func (p *HealthCheckFilter) FilterRecords(records []dns.RR) []dns.RR {
+	if p.cache.Len() == 0 {
```

Comment:
Do we really need this condition? Actually it's true only at the first request, after that cache won't be empty.